{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "mount_file_id": "12Mtvb92Jv3_AryL_ciIP7g37P-xgNgL2",
      "authorship_tag": "ABX9TyOovcgQ0PolR+vb5KAuPrlI",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/NehaSontakk/BATH-Prokka-Comparison/blob/main/phmmer_E_values.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": 19,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "T6-MlmV8_adh",
        "outputId": "8ed67208-fd26-4ffa-b58c-370b95ebf708"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Requirement already satisfied: biopython in /usr/local/lib/python3.10/dist-packages (1.84)\n",
            "Requirement already satisfied: numpy in /usr/local/lib/python3.10/dist-packages (from biopython) (1.26.4)\n"
          ]
        }
      ],
      "source": [
        "!pip install biopython"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "from Bio import SeqIO"
      ],
      "metadata": {
        "id": "Ttjb9oOm_jhP"
      },
      "execution_count": 20,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import re\n",
        "from Bio import SeqIO\n",
        "df = pd.read_excel(\"/content/drive/MyDrive/Lab Work/Parkinsons_Data/BIN152/Alignment_Overlaps_Unique_Bedmap_COGs_bin152.xlsx\")\n",
        "dna_fasta_file = \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/BIN152/Prokka Output/bin152.fna\""
      ],
      "metadata": {
        "id": "LXpjdAzsAHSO"
      },
      "execution_count": 21,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#Extract the information for each row of information\n",
        "def extract_info(row):\n",
        "    if pd.notnull(row['BATH']):\n",
        "        bath_data = re.split(r'\\t|\\s+', str(row['BATH']))\n",
        "        #print(bath_data)\n",
        "        row['BATH_contig'] = bath_data[0]\n",
        "        if len(bath_data) >= 3:\n",
        "            row['BATH_annot'] = bath_data[-3]\n",
        "            if int(bath_data[1]) < int(bath_data[2]):\n",
        "              row['BATH_start'] = abs(int(bath_data[1]))\n",
        "              row['BATH_end'] = abs(int(bath_data[2]))\n",
        "            else:\n",
        "              row['BATH_end'] = abs(int(bath_data[1]))\n",
        "              row['BATH_start'] = abs(int(bath_data[2]))\n",
        "        else:\n",
        "            row['BATH_annot'] = None\n",
        "\n",
        "    if pd.notnull(row['Prokka']):\n",
        "        prokka_data = re.split(r'\\t|\\s+', str(row['Prokka']))\n",
        "        prokka_annotr = re.split(r':', str(row['Prokka']))[-1]\n",
        "        row['Prokka_contig'] = prokka_data[0]\n",
        "        if len(prokka_data) >= 3:\n",
        "            #print(\"Prokka starts:\",int(prokka_data[1]))\n",
        "            row['Prokka_annot'] = prokka_annotr\n",
        "            if int(prokka_data[1]) < int(prokka_data[2]):\n",
        "              row['Prokka_start'] = abs(int(prokka_data[1]))\n",
        "              row['Prokka_end'] = abs(int(prokka_data[2]))\n",
        "            else:\n",
        "              row['Prokka_end'] = abs(int(prokka_data[1]) )\n",
        "              row['Prokka_start'] = abs(int(prokka_data[2]))\n",
        "        else:\n",
        "            row['Prokka_annot'] = None\n",
        "    return row"
      ],
      "metadata": {
        "id": "v07SOCBAAdti"
      },
      "execution_count": 22,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df2 = df.apply(extract_info, axis=1)"
      ],
      "metadata": {
        "id": "8A7o0rZbAgnR"
      },
      "execution_count": 23,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 524
        },
        "id": "vN_tUbEgSqzj",
        "outputId": "9cfc9b65-347d-4df0-dadf-e9f5b46c8f9c"
      },
      "execution_count": 24,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Unnamed: 0.1  Unnamed: 0                                        BATH  \\\n",
              "0              87          87       k127_2253129\\t901\\t2262\\tQ99SX1\\t+\\t0   \n",
              "1              90          90    k127_2253129\\t11353\\t11787\\tO35016\\t+\\t0   \n",
              "2              91          91       k127_2033476\\t541\\t1671\\tP80092\\t+\\t0   \n",
              "3              92          92    k127_2033476\\t3426\\t3785\\tMF_00101\\t+\\t0   \n",
              "4              93          93      k127_2033476\\t3829\\t5298\\tQ9X024\\t+\\t0   \n",
              "..            ...         ...                                         ...   \n",
              "292           401         401   k127_2982141\\t9733\\t11265\\tMF_00139\\t+\\t0   \n",
              "293           402         402  k127_2982141\\t11286\\t12548\\tMF_00138\\t+\\t0   \n",
              "294           403         403    k127_2982141\\t12931\\t13911\\tA3DDX7\\t+\\t0   \n",
              "295           405         405  k127_2982141\\t16989\\t17780\\tMF_00228\\t+\\t0   \n",
              "296           406         406  k127_2982141\\t17781\\t18393\\tMF_00097\\t+\\t0   \n",
              "\n",
              "    BATH_annot   BATH_contig  BATH_end  BATH_start  \\\n",
              "0       Q99SX1  k127_2253129      2262         901   \n",
              "1       O35016  k127_2253129     11787       11353   \n",
              "2       P80092  k127_2033476      1671         541   \n",
              "3     MF_00101  k127_2033476      3785        3426   \n",
              "4       Q9X024  k127_2033476      5298        3829   \n",
              "..         ...           ...       ...         ...   \n",
              "292   MF_00139  k127_2982141     11265        9733   \n",
              "293   MF_00138  k127_2982141     12548       11286   \n",
              "294     A3DDX7  k127_2982141     13911       12931   \n",
              "295   MF_00228  k127_2982141     17780       16989   \n",
              "296   MF_00097  k127_2982141     18393       17781   \n",
              "\n",
              "                                                Prokka Prokka_annot  \\\n",
              "0    k127_2253129\\t895\\t2301\\tJNLINIBG_00430\\t+\\tab...       Q7A4P8   \n",
              "1    k127_2253129\\t11320\\t11796\\tJNLINIBG_00436\\t+\\...       O35016   \n",
              "2    k127_2033476\\t517\\t1767\\tJNLINIBG_00225\\t+\\tab...       P54955   \n",
              "3    k127_2033476\\t3423\\t3815\\tJNLINIBG_00228\\t+\\ta...       Q9KPB6   \n",
              "4    k127_2033476\\t3796\\t5328\\tJNLINIBG_00229\\t+\\ta...       Q9X024   \n",
              "..                                                 ...          ...   \n",
              "292  k127_2982141\\t9727\\t11271\\tJNLINIBG_00634\\t+\\t...       P67544   \n",
              "293  k127_2982141\\t11286\\t12566\\tJNLINIBG_00635\\t+\\...       P12039   \n",
              "294  k127_2982141\\t12928\\t13914\\tJNLINIBG_00636\\t+\\...       A3DDX7   \n",
              "295  k127_2982141\\t16983\\t17789\\tJNLINIBG_00640\\t+\\...       P39593   \n",
              "296  k127_2982141\\t17776\\t18411\\tJNLINIBG_00641\\t+\\...       P39594   \n",
              "\n",
              "    Prokka_contig  ...  Prokka ORF HMM phmmer E-value  \\\n",
              "0    k127_2253129  ...                            NaN   \n",
              "1    k127_2253129  ...                            NaN   \n",
              "2    k127_2033476  ...                            NaN   \n",
              "3    k127_2033476  ...                            NaN   \n",
              "4    k127_2033476  ...                            NaN   \n",
              "..            ...  ...                            ...   \n",
              "292  k127_2982141  ...                            NaN   \n",
              "293  k127_2982141  ...                            NaN   \n",
              "294  k127_2982141  ...                            NaN   \n",
              "295  k127_2982141  ...                            NaN   \n",
              "296  k127_2982141  ...                            NaN   \n",
              "\n",
              "     Prokka ORF HMM phmmer ali from Prokka ORF HMM phmmer ali to  \\\n",
              "0                               NaN                          NaN   \n",
              "1                               NaN                          NaN   \n",
              "2                               NaN                          NaN   \n",
              "3                               NaN                          NaN   \n",
              "4                               NaN                          NaN   \n",
              "..                              ...                          ...   \n",
              "292                             NaN                          NaN   \n",
              "293                             NaN                          NaN   \n",
              "294                             NaN                          NaN   \n",
              "295                             NaN                          NaN   \n",
              "296                             NaN                          NaN   \n",
              "\n",
              "    Prokka ORF HMM phmmer length Merged_Column  Prokka_SW  Annotation_Match  \\\n",
              "0                            NaN           471      Blast                No   \n",
              "1                            NaN           145      Blast               Yes   \n",
              "2                            NaN           354      Blast                No   \n",
              "3                            NaN           118      Blast                No   \n",
              "4                            NaN           480      Blast               Yes   \n",
              "..                           ...           ...        ...               ...   \n",
              "292                          NaN           490      Blast                No   \n",
              "293                          NaN           417      Blast                No   \n",
              "294                          NaN           325      Blast               Yes   \n",
              "295                          NaN           256      Blast                No   \n",
              "296                          NaN           184      Blast                No   \n",
              "\n",
              "    Prokka COGs   BATH COGs Name_or_COG_Match  \n",
              "0           NaN         NaN                No  \n",
              "1       COG0394     COG0394               Yes  \n",
              "2       COG1473  arCOG01108                No  \n",
              "3       COG0736     COG0736               Yes  \n",
              "4       COG0062     COG0062               Yes  \n",
              "..          ...         ...               ...  \n",
              "292         NaN     KOG2555                No  \n",
              "293     COG0151     KOG0237                No  \n",
              "294     COG1748     COG1748               Yes  \n",
              "295     COG2145     COG2145               Yes  \n",
              "296     COG0352     KOG2598                No  \n",
              "\n",
              "[297 rows x 38 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-ba70f23c-f2b1-4c14-bfda-26906e37896a\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Unnamed: 0.1</th>\n",
              "      <th>Unnamed: 0</th>\n",
              "      <th>BATH</th>\n",
              "      <th>BATH_annot</th>\n",
              "      <th>BATH_contig</th>\n",
              "      <th>BATH_end</th>\n",
              "      <th>BATH_start</th>\n",
              "      <th>Prokka</th>\n",
              "      <th>Prokka_annot</th>\n",
              "      <th>Prokka_contig</th>\n",
              "      <th>...</th>\n",
              "      <th>Prokka ORF HMM phmmer E-value</th>\n",
              "      <th>Prokka ORF HMM phmmer ali from</th>\n",
              "      <th>Prokka ORF HMM phmmer ali to</th>\n",
              "      <th>Prokka ORF HMM phmmer length</th>\n",
              "      <th>Merged_Column</th>\n",
              "      <th>Prokka_SW</th>\n",
              "      <th>Annotation_Match</th>\n",
              "      <th>Prokka COGs</th>\n",
              "      <th>BATH COGs</th>\n",
              "      <th>Name_or_COG_Match</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>87</td>\n",
              "      <td>87</td>\n",
              "      <td>k127_2253129\\t901\\t2262\\tQ99SX1\\t+\\t0</td>\n",
              "      <td>Q99SX1</td>\n",
              "      <td>k127_2253129</td>\n",
              "      <td>2262</td>\n",
              "      <td>901</td>\n",
              "      <td>k127_2253129\\t895\\t2301\\tJNLINIBG_00430\\t+\\tab...</td>\n",
              "      <td>Q7A4P8</td>\n",
              "      <td>k127_2253129</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>471</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>90</td>\n",
              "      <td>90</td>\n",
              "      <td>k127_2253129\\t11353\\t11787\\tO35016\\t+\\t0</td>\n",
              "      <td>O35016</td>\n",
              "      <td>k127_2253129</td>\n",
              "      <td>11787</td>\n",
              "      <td>11353</td>\n",
              "      <td>k127_2253129\\t11320\\t11796\\tJNLINIBG_00436\\t+\\...</td>\n",
              "      <td>O35016</td>\n",
              "      <td>k127_2253129</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>145</td>\n",
              "      <td>Blast</td>\n",
              "      <td>Yes</td>\n",
              "      <td>COG0394</td>\n",
              "      <td>COG0394</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>91</td>\n",
              "      <td>91</td>\n",
              "      <td>k127_2033476\\t541\\t1671\\tP80092\\t+\\t0</td>\n",
              "      <td>P80092</td>\n",
              "      <td>k127_2033476</td>\n",
              "      <td>1671</td>\n",
              "      <td>541</td>\n",
              "      <td>k127_2033476\\t517\\t1767\\tJNLINIBG_00225\\t+\\tab...</td>\n",
              "      <td>P54955</td>\n",
              "      <td>k127_2033476</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>354</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>COG1473</td>\n",
              "      <td>arCOG01108</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>92</td>\n",
              "      <td>92</td>\n",
              "      <td>k127_2033476\\t3426\\t3785\\tMF_00101\\t+\\t0</td>\n",
              "      <td>MF_00101</td>\n",
              "      <td>k127_2033476</td>\n",
              "      <td>3785</td>\n",
              "      <td>3426</td>\n",
              "      <td>k127_2033476\\t3423\\t3815\\tJNLINIBG_00228\\t+\\ta...</td>\n",
              "      <td>Q9KPB6</td>\n",
              "      <td>k127_2033476</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>118</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>COG0736</td>\n",
              "      <td>COG0736</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>93</td>\n",
              "      <td>93</td>\n",
              "      <td>k127_2033476\\t3829\\t5298\\tQ9X024\\t+\\t0</td>\n",
              "      <td>Q9X024</td>\n",
              "      <td>k127_2033476</td>\n",
              "      <td>5298</td>\n",
              "      <td>3829</td>\n",
              "      <td>k127_2033476\\t3796\\t5328\\tJNLINIBG_00229\\t+\\ta...</td>\n",
              "      <td>Q9X024</td>\n",
              "      <td>k127_2033476</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>480</td>\n",
              "      <td>Blast</td>\n",
              "      <td>Yes</td>\n",
              "      <td>COG0062</td>\n",
              "      <td>COG0062</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>292</th>\n",
              "      <td>401</td>\n",
              "      <td>401</td>\n",
              "      <td>k127_2982141\\t9733\\t11265\\tMF_00139\\t+\\t0</td>\n",
              "      <td>MF_00139</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>11265</td>\n",
              "      <td>9733</td>\n",
              "      <td>k127_2982141\\t9727\\t11271\\tJNLINIBG_00634\\t+\\t...</td>\n",
              "      <td>P67544</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>490</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>NaN</td>\n",
              "      <td>KOG2555</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>293</th>\n",
              "      <td>402</td>\n",
              "      <td>402</td>\n",
              "      <td>k127_2982141\\t11286\\t12548\\tMF_00138\\t+\\t0</td>\n",
              "      <td>MF_00138</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>12548</td>\n",
              "      <td>11286</td>\n",
              "      <td>k127_2982141\\t11286\\t12566\\tJNLINIBG_00635\\t+\\...</td>\n",
              "      <td>P12039</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>417</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>COG0151</td>\n",
              "      <td>KOG0237</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>294</th>\n",
              "      <td>403</td>\n",
              "      <td>403</td>\n",
              "      <td>k127_2982141\\t12931\\t13911\\tA3DDX7\\t+\\t0</td>\n",
              "      <td>A3DDX7</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>13911</td>\n",
              "      <td>12931</td>\n",
              "      <td>k127_2982141\\t12928\\t13914\\tJNLINIBG_00636\\t+\\...</td>\n",
              "      <td>A3DDX7</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>325</td>\n",
              "      <td>Blast</td>\n",
              "      <td>Yes</td>\n",
              "      <td>COG1748</td>\n",
              "      <td>COG1748</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>295</th>\n",
              "      <td>405</td>\n",
              "      <td>405</td>\n",
              "      <td>k127_2982141\\t16989\\t17780\\tMF_00228\\t+\\t0</td>\n",
              "      <td>MF_00228</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>17780</td>\n",
              "      <td>16989</td>\n",
              "      <td>k127_2982141\\t16983\\t17789\\tJNLINIBG_00640\\t+\\...</td>\n",
              "      <td>P39593</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>256</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>COG2145</td>\n",
              "      <td>COG2145</td>\n",
              "      <td>Yes</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>296</th>\n",
              "      <td>406</td>\n",
              "      <td>406</td>\n",
              "      <td>k127_2982141\\t17781\\t18393\\tMF_00097\\t+\\t0</td>\n",
              "      <td>MF_00097</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>18393</td>\n",
              "      <td>17781</td>\n",
              "      <td>k127_2982141\\t17776\\t18411\\tJNLINIBG_00641\\t+\\...</td>\n",
              "      <td>P39594</td>\n",
              "      <td>k127_2982141</td>\n",
              "      <td>...</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>184</td>\n",
              "      <td>Blast</td>\n",
              "      <td>No</td>\n",
              "      <td>COG0352</td>\n",
              "      <td>KOG2598</td>\n",
              "      <td>No</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>297 rows × 38 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ba70f23c-f2b1-4c14-bfda-26906e37896a')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-ba70f23c-f2b1-4c14-bfda-26906e37896a button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-ba70f23c-f2b1-4c14-bfda-26906e37896a');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "<div id=\"df-9132a30e-ece3-455d-9644-aff2fee154d9\">\n",
              "  <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-9132a30e-ece3-455d-9644-aff2fee154d9')\"\n",
              "            title=\"Suggest charts\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "  </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "  <script>\n",
              "    async function quickchart(key) {\n",
              "      const quickchartButtonEl =\n",
              "        document.querySelector('#' + key + ' button');\n",
              "      quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "      quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "      try {\n",
              "        const charts = await google.colab.kernel.invokeFunction(\n",
              "            'suggestCharts', [key], {});\n",
              "      } catch (error) {\n",
              "        console.error('Error during call to suggestCharts:', error);\n",
              "      }\n",
              "      quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "      quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "    }\n",
              "    (() => {\n",
              "      let quickchartButtonEl =\n",
              "        document.querySelector('#df-9132a30e-ece3-455d-9644-aff2fee154d9 button');\n",
              "      quickchartButtonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "    })();\n",
              "  </script>\n",
              "</div>\n",
              "\n",
              "  <div id=\"id_1b0b992c-ba28-4db7-b3de-513b3162482b\">\n",
              "    <style>\n",
              "      .colab-df-generate {\n",
              "        background-color: #E8F0FE;\n",
              "        border: none;\n",
              "        border-radius: 50%;\n",
              "        cursor: pointer;\n",
              "        display: none;\n",
              "        fill: #1967D2;\n",
              "        height: 32px;\n",
              "        padding: 0 0 0 0;\n",
              "        width: 32px;\n",
              "      }\n",
              "\n",
              "      .colab-df-generate:hover {\n",
              "        background-color: #E2EBFA;\n",
              "        box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "        fill: #174EA6;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate {\n",
              "        background-color: #3B4455;\n",
              "        fill: #D2E3FC;\n",
              "      }\n",
              "\n",
              "      [theme=dark] .colab-df-generate:hover {\n",
              "        background-color: #434B5C;\n",
              "        box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "        filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "        fill: #FFFFFF;\n",
              "      }\n",
              "    </style>\n",
              "    <button class=\"colab-df-generate\" onclick=\"generateWithVariable('df')\"\n",
              "            title=\"Generate code using this dataframe.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M7,19H8.4L18.45,9,17,7.55,7,17.6ZM5,21V16.75L18.45,3.32a2,2,0,0,1,2.83,0l1.4,1.43a1.91,1.91,0,0,1,.58,1.4,1.91,1.91,0,0,1-.58,1.4L9.25,21ZM18.45,9,17,7.55Zm-12,3A5.31,5.31,0,0,0,4.9,8.1,5.31,5.31,0,0,0,1,6.5,5.31,5.31,0,0,0,4.9,4.9,5.31,5.31,0,0,0,6.5,1,5.31,5.31,0,0,0,8.1,4.9,5.31,5.31,0,0,0,12,6.5,5.46,5.46,0,0,0,6.5,12Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "    <script>\n",
              "      (() => {\n",
              "      const buttonEl =\n",
              "        document.querySelector('#id_1b0b992c-ba28-4db7-b3de-513b3162482b button.colab-df-generate');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      buttonEl.onclick = () => {\n",
              "        google.colab.notebook.generateWithVariable('df');\n",
              "      }\n",
              "      })();\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "variable_name": "df"
            }
          },
          "metadata": {},
          "execution_count": 24
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Get BATH E-values"
      ],
      "metadata": {
        "id": "VSz1jRDlCZLC"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "bath_protein = pd.read_excel(\"/content/drive/MyDrive/Lab Work/Parkinsons_Data/BIN152/Deduplication Output/dedup_bin52_test1.xlsx\")\n",
        "unnamed_cols = [col for col in bath_protein.columns if col.startswith('Unnamed')]\n",
        "bath_protein.drop(columns=unnamed_cols, inplace=True)"
      ],
      "metadata": {
        "id": "aXfRZR2YBOz0"
      },
      "execution_count": 25,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3 = pd.merge(df2,bath_protein[['target name','query name','E-value','shifts']],how=\"left\",left_on=['BATH_contig','BATH_annot'],right_on=['target name','query name'])"
      ],
      "metadata": {
        "id": "aYq1myE-FLxn"
      },
      "execution_count": 26,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "!sudo apt-get update\n",
        "!sudo apt-get install hmmer"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "c2524137-9d26-46f1-e9f9-38230584b2b9",
        "id": "Nax2Uzr2ky7D"
      },
      "execution_count": 27,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "\r0% [Working]\r            \rHit:1 http://security.ubuntu.com/ubuntu jammy-security InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com (185.125.190.82)] [Connected to cloud.r-pr\r                                                                               \rHit:2 https://cloud.r-project.org/bin/linux/ubuntu jammy-cran40/ InRelease\n",
            "\r0% [Connecting to archive.ubuntu.com (185.125.190.82)] [Connected to r2u.stat.i\r                                                                               \rHit:3 https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64  InRelease\n",
            "\r0% [Waiting for headers] [Waiting for headers] [Connecting to ppa.launchpadcont\r                                                                               \rIgn:4 https://r2u.stat.illinois.edu/ubuntu jammy InRelease\n",
            "\r0% [Waiting for headers] [Connected to ppa.launchpadcontent.net (185.125.190.80\r                                                                               \rHit:5 http://archive.ubuntu.com/ubuntu jammy InRelease\n",
            "Hit:6 https://r2u.stat.illinois.edu/ubuntu jammy Release\n",
            "Hit:7 http://archive.ubuntu.com/ubuntu jammy-updates InRelease\n",
            "Hit:8 http://archive.ubuntu.com/ubuntu jammy-backports InRelease\n",
            "Hit:9 https://ppa.launchpadcontent.net/deadsnakes/ppa/ubuntu jammy InRelease\n",
            "Hit:11 https://ppa.launchpadcontent.net/graphics-drivers/ppa/ubuntu jammy InRelease\n",
            "Hit:12 https://ppa.launchpadcontent.net/ubuntugis/ppa/ubuntu jammy InRelease\n",
            "Reading package lists... Done\n",
            "W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)\n",
            "Reading package lists... Done\n",
            "Building dependency tree... Done\n",
            "Reading state information... Done\n",
            "hmmer is already the newest version (3.3.2+dfsg-1).\n",
            "0 upgraded, 0 newly installed, 0 to remove and 45 not upgraded.\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Finding the Prokka protein\n",
        "import os\n",
        "\n",
        "file_paths = [\n",
        "    \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/kingdom/Bacteria/sprot\",\n",
        "    \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/kingdom/Bacteria/IS\",\n",
        "    \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/kingdom/Bacteria/AMR\",\n",
        "    \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/kingdom/Archaea/sprot\",\n",
        "    \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/kingdom/Mitochondria/sprot\",\n",
        "    \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/kingdom/Viruses/sprot\"\n",
        "]\n",
        "combined_file = \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/combined_sequences.fasta\"\n",
        "\n",
        "with open(combined_file, 'w') as combined_fasta:\n",
        "    for file_path in file_paths:\n",
        "        if os.path.exists(file_path):\n",
        "            for record in SeqIO.parse(file_path, \"fasta\"):\n",
        "                SeqIO.write(record, combined_fasta, \"fasta\")\n",
        "        else:\n",
        "            print(f\"File not found: {file_path}\")\n",
        "\n",
        "def get_fasta_entries(fasta_path, protein_name):\n",
        "    #print(protein_name)\n",
        "    if pd.isna(protein_name):\n",
        "          return None\n",
        "    for record in SeqIO.parse(fasta_path, \"fasta\"):\n",
        "        if protein_name == record.name:\n",
        "          #print(record.seq)\n",
        "          return str(record.seq)\n",
        "\n",
        "def get_prokka_fasta_sequence(row):\n",
        "    return get_fasta_entries(combined_file, row['Prokka_annot'])\n",
        "\n"
      ],
      "metadata": {
        "id": "gnKWdvdLLJ5Q"
      },
      "execution_count": 28,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import numpy as np\n",
        "df3['Prokka_annot'] = df3['Prokka_annot'].str.replace(\"002006\", \"\", regex=False)\n",
        "df3['Prokka_annot'].replace(\"\", np.nan, inplace=True)"
      ],
      "metadata": {
        "id": "h3StJrIiSDh-"
      },
      "execution_count": 29,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3['Prokka Uniprot Protein Sequence'] = df3.apply(get_prokka_fasta_sequence,axis=1)"
      ],
      "metadata": {
        "id": "02-f3NWmL14H"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3[~df3['Prokka_annot'].isna()]['Prokka Uniprot Protein Sequence']"
      ],
      "metadata": {
        "id": "QnhKMmXkQpSM"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3"
      ],
      "metadata": {
        "id": "xjwi8D7TR_gB"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "def get_bath_fasta_sequence(row):\n",
        "    return get_fasta_entries(combined_file, row['BATH_annot'])"
      ],
      "metadata": {
        "id": "0YbA7ZJIQu_V"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3['BATH Uniprot Protein Sequence'] = df3.apply(get_bath_fasta_sequence,axis=1)"
      ],
      "metadata": {
        "id": "A36lIhLSQ84C"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import subprocess\n",
        "\n",
        "def fetch_onlyhmm_path(protein_name, hmm_db_path):\n",
        "    hmm_output_path = f\"/content/output_{protein_name}.hmm\"\n",
        "    fetch_command = f\"hmmfetch -o '{hmm_output_path}' '{hmm_db_path}' '{protein_name}'\"\n",
        "    result = os.system(fetch_command)\n",
        "    if result != 0:\n",
        "        print(f\"Failed to fetch HMM for {protein_name} from {hmm_db_path}\")\n",
        "        return None\n",
        "    if os.path.getsize(hmm_output_path) == 0:  # Check if the file is empty\n",
        "        print(f\"Warning: Generated HMM file {hmm_output_path} is empty.\")\n",
        "        return None\n",
        "    return hmm_output_path\n",
        "\n",
        "\n",
        "\n",
        "def run_phmmer_HAMAP_Uniprot(row,output_dir):\n",
        "  out_path = None\n",
        "  result = None\n",
        "  if pd.isna(row['Prokka_annot']) or pd.isna(row['BATH_annot']):\n",
        "        return np.nan, np.nan, np.nan\n",
        "  print(row['Prokka_annot'],row['BATH_annot'])\n",
        "  if row['Prokka_annot'].startswith('MF_') or row['BATH_annot'].startswith('MF_'):\n",
        "      if row['Prokka_annot'].startswith('MF_'):\n",
        "        print(row['BATH Uniprot Protein Sequence'])\n",
        "        out_path = os.path.join(output_dir, f\"{row['Prokka_annot']}_{row['BATH_annot']}_results.out\")\n",
        "        #Fetch the HAMAP hmm file for this\n",
        "        prokka_hmm_output_path = fetch_onlyhmm_path(row['Prokka_annot'], hmm_db_path)\n",
        "        print(prokka_hmm_output_path)\n",
        "        #Make a second protein file for the bath sequence\n",
        "        protein_sequence_path = os.path.join(output_dir, f\"{row['BATH_annot']}_forphmmer.fa\")\n",
        "        with open(protein_sequence_path, 'w') as protein_file:\n",
        "            protein_file.write(f\">{row['BATH_annot']}\\n{row['BATH Uniprot Protein Sequence']}\\n\")\n",
        "        cmd = ['phmmer', '--tblout', out_path, protein_sequence_path, prokka_hmm_output_path]\n",
        "        result = subprocess.run(cmd, capture_output=True, text=True)\n",
        "        #os.remove(protein_sequence_path)\n",
        "        #os.remove(prokka_hmm_output_path)\n",
        "\n",
        "        if result.returncode != 0:\n",
        "          return np.nan, np.nan, np.nan\n",
        "        return result.stdout\n",
        "\n",
        "      elif row['BATH_annot'].startswith('MF_'):\n",
        "        #Fetch the HAMAP file for BATH\n",
        "        bath_hmm_output_path = fetch_onlyhmm_path(row['BATH_annot'], hmm_db_path)\n",
        "        out_path = os.path.join(output_dir, f\"{row['Prokka_annot']}_{row['BATH_annot']}_results.out\")\n",
        "        protein_sequence_path = os.path.join(output_dir, f\"{row['Prokka_annot']}_forphmmer.fa\")\n",
        "        with open(protein_sequence_path, 'w') as protein_file:\n",
        "            protein_file.write(f\">{row['Prokka_annot']}\\n{row.get('Prokka Uniprot Protein Sequence', '')}\\n\")\n",
        "        cmd = ['phmmer', '--tblout', out_path, protein_sequence_path, bath_hmm_output_path]\n",
        "        result = subprocess.run(cmd, capture_output=True, text=True)\n",
        "        os.remove(protein_sequence_path)\n",
        "        os.remove(prokka_hmm_output_path)\n",
        "\n",
        "\n",
        "  if row['Prokka_annot'].startswith('MF_') and row['BATH_annot'].startswith('MF_'):\n",
        "      #print(\"Two MFs\")\n",
        "      out_path = os.path.join(output_dir, f\"{row['Prokka_annot']}_{row['BATH_annot']}_results.out\")\n",
        "      bath_hmm_output_path = fetch_onlyhmm_path(row['BATH_annot'], hmm_db_path)\n",
        "      prokka_hmm_output_path = fetch_onlyhmm_path(row['Prokka_annot'], hmm_db_path)\n",
        "      cmd = ['phmmer', '--tblout', out_path, bath_hmm_output_path, prokka_hmm_output_path]\n",
        "      result = subprocess.run(cmd, capture_output=True, text=True)\n",
        "      os.remove(protein_sequence_path)\n",
        "\n",
        "  if not row['Prokka_annot'].startswith('MF_') or not row['BATH_annot'].startswith('MF_'):\n",
        "      out_path = os.path.join(output_dir, f\"{row['Prokka_annot']}_{row['BATH_annot']}_results.out\")\n",
        "      prokka_protein_sequence_path = os.path.join(output_dir, f\"{row['Prokka_annot']}_forphmmer.fa\")\n",
        "      with open(prokka_protein_sequence_path, 'w') as protein_file:\n",
        "          protein_file.write(f\">{row['Prokka_annot']}\\n{row.get('Prokka Uniprot Protein Sequence', '')}\\n\")\n",
        "      bath_protein_sequence_path = os.path.join(output_dir, f\"{row['BATH_annot']}_forphmmer.fa\")\n",
        "      with open(bath_protein_sequence_path, 'w') as protein_file:\n",
        "          protein_file.write(f\">{row['BATH_annot']}\\n{row.get('BATH Uniprot Protein Sequence', '')}\\n\")\n",
        "      cmd = ['phmmer', '--tblout', out_path, bath_protein_sequence_path, prokka_protein_sequence_path]\n",
        "      result = subprocess.run(cmd, capture_output=True, text=True)\n",
        "      try:\n",
        "\n",
        "        if os.path.exists(prokka_protein_sequence_path):\n",
        "            os.remove(prokka_protein_sequence_path)\n",
        "        if os.path.exists(bath_protein_sequence_path):\n",
        "            os.remove(bath_protein_sequence_path)\n",
        "      except Exception as e:\n",
        "          print(f\"Error removing files: {e}\")\n",
        "\n",
        "  e_value, i_evalue, c_evalue = \"No match\", \"No match\", \"No match\"\n",
        "  found_match = False\n",
        "\n",
        "  if result and result.returncode == 0:\n",
        "      with open(out_path, 'r') as file:\n",
        "          for line in file:\n",
        "              if not line.startswith(\"#\"):\n",
        "                  found_match = True\n",
        "                  parts = line.strip().split()\n",
        "                  if len(parts) > 11:  # Verify the column count\n",
        "                      e_value = parts[4]\n",
        "                      #print(e_value)\n",
        "                      i_evalue = parts[11]  # Typical i-Evalue column\n",
        "                      c_evalue = parts[10]  # Typical c-Evalue column\n",
        "                      os.remove(out_path)\n",
        "                      break\n",
        "  elif result:\n",
        "      print(f\"Error running phmmer: {result.stderr}\")\n",
        "\n",
        "  if not found_match:\n",
        "      os.remove(out_path)\n",
        "      return \"No significant matches found\", \"No significant matches found\", \"No significant matches found\"\n",
        "\n",
        "  return e_value, i_evalue, c_evalue\n",
        "\n",
        "\n",
        "\n",
        "output_dir = \"/content/\"\n",
        "#df3.apply(lambda row: run_phmmer_HAMAP_Uniprot(row), axis=1)\n",
        "hmm_db_path = \"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Prokka db/hmm/HAMAP.hmm\"\n",
        "df3['phmmer_eic_values'] = df3.apply(run_phmmer_HAMAP_Uniprot, axis=1, args=(output_dir,))"
      ],
      "metadata": {
        "id": "1tKOc6yYkq4u"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3['phmmer_eic_values'].dropna()"
      ],
      "metadata": {
        "id": "7D_VKI5kfYrv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3[['phmmeroutput_e_value', 'phmmeroutput_i_value', 'phmmeroutput_c_value']] = pd.DataFrame(df3['phmmer_eic_values'].tolist(), index=df3.index)"
      ],
      "metadata": {
        "id": "jdUeez1OglJW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3[['phmmeroutput_e_value', 'phmmeroutput_i_value', 'phmmeroutput_c_value','Name_or_COG_Match']].dropna()"
      ],
      "metadata": {
        "id": "qhskhbuQm1Jx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3.to_excel(\"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Organized/phmmer_Evalues_COG_Matches.xlsx\")"
      ],
      "metadata": {
        "id": "zh7c_PdCs0R4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import pandas as pd\n",
        "import numpy as np\n",
        "import matplotlib.pyplot as plt\n",
        "import seaborn as sns\n",
        "\n",
        "plt.style.use('default')\n",
        "\n",
        "# Load the data\n",
        "df3 = pd.read_excel(\"/content/drive/MyDrive/Lab Work/Parkinsons_Data/Organized/phmmer_Evalues_COG_Matches.xlsx\")\n",
        "\n",
        "# Filter out non-numeric values\n",
        "numeric_values = pd.to_numeric(df3['phmmeroutput_e_value'], errors='coerce')\n",
        "\n",
        "# Apply log normalization only to numeric values\n",
        "df3['Log_phmmeroutput_e_value'] = np.log10(numeric_values)\n",
        "\n",
        "# Melt the DataFrame for plotting\n",
        "melted_df = pd.melt(df3, id_vars='Name_or_COG_Match', value_vars=['Log_phmmeroutput_e_value'],\n",
        "                    var_name='Variable', value_name='Log_Evalue')\n",
        "\n",
        "# Define the color palette\n",
        "palette = ['#448361', '#D44C47']\n",
        "\n",
        "# Set font size\n",
        "font_size = 20\n",
        "plt.rc('font', size=font_size)          # controls default text sizes\n",
        "plt.rc('axes', titlesize=font_size)     # fontsize of the axes title\n",
        "plt.rc('axes', labelsize=font_size)     # fontsize of the x and y labels\n",
        "plt.rc('xtick', labelsize=font_size)    # fontsize of the tick labels\n",
        "plt.rc('ytick', labelsize=font_size)    # fontsize of the tick labels\n",
        "plt.rc('legend', fontsize=font_size)    # legend fontsize\n",
        "plt.rc('figure', titlesize=font_size)   # fontsize of the figure title\n",
        "\n",
        "# Plotting the violin plot along with strip plot for detailed data view\n",
        "plt.figure(figsize=(12, 7))\n",
        "sns.violinplot(x='Name_or_COG_Match', y='Log_Evalue', data=melted_df, palette=palette)\n",
        "sns.stripplot(x='Name_or_COG_Match', y='Log_Evalue', data=melted_df, edgecolor='black', linewidth=0.5, color='grey', alpha=0.7, jitter=True)\n",
        "\n",
        "# Customize the x-axis labels\n",
        "custom_labels = ['Mismatching Annotations\\nbetween BATH and Prokka', 'Matching Annotations\\nbetween BATH and Prokka']\n",
        "plt.xticks(ticks=range(len(custom_labels)), labels=custom_labels, rotation=0, ha='center')\n",
        "\n",
        "# Calculate and annotate the median and quartile values\n",
        "for idx, label in enumerate(melted_df['Name_or_COG_Match'].unique()):\n",
        "    label_data = melted_df[melted_df['Name_or_COG_Match'] == label]['Log_Evalue']\n",
        "    median = label_data.median()\n",
        "    q1 = label_data.quantile(0.25)\n",
        "    q3 = label_data.quantile(0.75)\n",
        "\n",
        "    plt.text(idx, median, f'{median:.1f}', ha='center', va='top', color='white', fontweight='bold')\n",
        "    plt.plot([idx-0.2, idx+0.2], [median, median], color='black', lw=2)\n",
        "    plt.plot([idx-0.1, idx+0.1], [q1, q1], color='black', lw=1, linestyle='--')\n",
        "    plt.plot([idx-0.1, idx+0.1], [q3, q3], color='black', lw=1, linestyle='--')\n",
        "    plt.plot([idx, idx], [q1, q3], color='black', lw=1, linestyle='--')\n",
        "\n",
        "# Customize the plot appearance\n",
        "ax = plt.gca()\n",
        "ax.spines['top'].set_visible(False)\n",
        "ax.spines['right'].set_visible(False)\n",
        "ax.spines['bottom'].set_visible(True)\n",
        "ax.spines['left'].set_visible(True)\n",
        "\n",
        "ylim = plt.gca().get_ylim()\n",
        "yticks = np.linspace(ylim[0], ylim[1], num=10)  # Increase the number of intervals\n",
        "plt.yticks(yticks)\n",
        "\n",
        "# Set title and labels\n",
        "#plt.title('pHmmer E-Values for matching and mismatching annotations from BATH and Prokka')\n",
        "plt.xlabel('')\n",
        "plt.ylabel('E-values from pHMMER\\nprotein alignments (Log10)')\n",
        "plt.tight_layout()\n",
        "plt.savefig(\"Fig5B_phmmerProkkaBATH.svg\", dpi=300, transparent=True, bbox_inches='tight')\n",
        "plt.show()\n"
      ],
      "metadata": {
        "id": "wHIn5xfRmUbb"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3['Prokka'].dropna()"
      ],
      "metadata": {
        "id": "y6RQTZ5pKjWV"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3['BATH'].dropna()"
      ],
      "metadata": {
        "id": "xzdVQNp0LHZv"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3['Annotation_Match'].value_counts()"
      ],
      "metadata": {
        "id": "vvZZxQBxLmsc"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "df3.columns"
      ],
      "metadata": {
        "id": "9FTdjm2hM9vh"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}